Add executor for v8

VadimPlh opened the pull request:
## Cover letter

Describe in plain language the motivation (bug, feature, etc.) behind the change in this PR and how the included commits address it.

Fixes: #NNN, #NNN, ...

## Release notes

If the PR changes the user experience, write a short summary of the changes. See the [CONTRIBUTING](https://github.com/vectorizedio/redpanda/blob/dev/CONTRIBUTING.md) guidelines for details.

Release note: [1-2 sentences of what this PR changes]

0x5d committed:
tests: Update the rpk clients to use the topic namespace

0x5d committed:
tests: Rename rpk_test to rpk_topic_test

Since there will be more rpk tests for the rest of the commands,
it's better to separate them into different files.

Commit:
tests/ducky: Add wrapper function for rpk cluster info

Commit:
tests/ducky: Add test suite for rpk cluster*

Commit:
tests/rpk/topic: Fix existing tests

Commit:
tests: Add remote rpk client

The existing ducktape rpk client (RpkTool) executes commands on
the ducktape controller container. However, config changes are done
on the nodes themselves, so this commit introduces a ducktape rpk
client which runs commands on the given node.

Commit:
tests/rpk: Add tests for rpk config

Commit:
rptests: Add a lock to restrict access to the config file

In the rpk produce/ consume tests, an async consumer service is
used (RpkConsumer, in services/rpk_consumer.py), they both use
RedpandaService#brokers, which opens and loads the nodes'
redpanda.yaml config file.

In 99.99999% of the cases, this creates a deadlock and the tests
hang, which is solved by locking access to the config file.

Commit:
tests/rpk: Default stdin to None

Commit:
tests: Capture stdout in the rpk client

rpk was recently changed to use stdout for logging, instead of
stderr, so change RpkClient to capture stdout.

twmb committed:
add package testfs

This is (currently) a small package that shoule help make setting up
filesystems in tests easier. I took a look at a few tests and noticed
they all do a bunch of linear setup, I think some of this could be
replaced in the future with a simpler map.

twmb committed:
rpk: add autocompletion to plugins

This is a bit less straightforward than just simple rpk autocompletion.

Rather than only figure out what plugins are available when we see args
that rpk does not understand, we can list all plugins always on startup.
These can be used for autocompletion.

The way that autocompletion works within cobra is that all commands and
their subcommands get dedicated functions within the generated shell
scripts. So, to get autocompletion of our listed plugins, we need our
plugins to be listed in the shell scripts, and to do that, we have to
add a whole bunch of tiny cobra.Commands corresponding to our plugins
before rpk executes the "generate shell-autocomplete" command.

Lastly, as one extra point of improvement, this defines "rpk.ac-" to be
the opt-in prefix to use if a plugin supports the --help-autocomplete
flag. If a plugin supports this, then it can return the use, long,
short, and valid args that we can use in our tiny cobra.Commands to make
them actually useful when running `rpk foo bar --help`. Basically, we
get the help text from the plugin, and we get its autocomplete args.

I've tested this all locally with a few scripts printing hello, and one
Go program that supports --help-autocomplete that returns

    ext
    ext-foo
    ext-foo-bat
    ext-bar

BenPope committed:
schema_registry: Introduce compatibility_level

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/store: Introduce compatibility

Introduce the global and subject compatibility interface.

* Set and get global compatibility.
* Set and get subject compatibility.
* If a subject is not found, return subject_not_found
* If a subject does not have a specific value, fall back to global

Signed-off-by: Ben Pope <ben@vectorized.io>

Commit:
ducktape: Fix async rpk consumer

Update the command to use the new "topic" namespace.

Sometimes, the consumer might start before the cluster is ready (e.g.
it's in the middle of an election), so add a retry loop to account
for that.

Override clean_node to prevent warning messages from being printed
each time the consumer is started.

BenPope committed:
schema_registry/storage: Introduce serde tests

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/storage: Introduce operator<< for schema_type

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/storage: Introduce operators for schema_key

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/storage: Introduce operators for schema_value

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/storage: Introduce config_value

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
pandaproxy/json: Introduce minify, prettify

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/requests: Introduce put_config

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Introduce put_config handler

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Wire-up put_config handler

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Introduce swagger for get_config

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/requests: Introduce get_config

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Introduce get_config handler

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Wire-up get_config handler

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Introduce swagger for put_config_subject

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Introduce put_config_subject handler

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Wire-up put_config_subject handler

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Introduce swagger for get_config_subject

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Introduce get_config_subject handler

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Wire-up get_config_subject handler

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/storage: Refactoring for future changes

* Make json <-> iobuf more generic
* Split consume_to_store to allow more keytypes

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/storage: Introduce config_key

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/storage: Introduce topic_key_type_handler

This is used to distinguish keytypes.

It's slightly different to most handlers, since it's very permissive
of unknown values. It only looks for {"keytype":"<keytype>"}

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/storage: Support config in consume_to_store

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/storage: Introduce make_config_batch

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/store: Switch to subject_entry

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/ducktape: Smoketest config endpoints

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Introduce from_string_view<schema_type>

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/storage: Make topic_key_type an enum

Signed-off-by: Ben Pope <ben@vectorized.io>

Commit:
operator: Create feature flag that removes trailing dot from DNS name

The Java requires that dns names does not have trailing dot.

Fixes #1640

Commit:
operator: Remove unused const

BenPope committed:
schema_registry: Introduce exceptions.h

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Introduce error_code::topic_parse_error

reply_error_code::zookeeper_error is poorly named, but the value is correct
for the return body.

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/storage: Futurize consume_to_store

Report errors via an ss::future<>

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry/store: Introduce upsert

Signed-off-by: Ben Pope <ben@vectorized.io>

Commit:
operator: Create jks and pkcs#12 in certificate secret

Fixes: #1661

Commit:
operator/e2e: Verify that schema registry works with server side TLS

Commit:
operator: Prevent nil pointer panic

zimbatm committed:
operator/helm: sync configurator image tag with operator tag

The configurator was always using the "latest" tag, which is bad for
reproducibility.

Ideally, the configurator image and tag would be independently
configurable, but using the same tag version seems like a good
compromise in the meantime.

RafalKorepta committed:
Merge pull request #1688 from RafalKorepta/rk/fix/GH-1640/remove-trailing-dot

Give feature toggle that removes trailing dot from advertised kafka api

mmaslankaprv committed:
r/errc: introduced error code for not existing group

Signed-off-by: Michal Maslanka <michal@vectorized.io>

mmaslankaprv committed:
r/consensus: do not log uninitialized target node id

When node id is not initialized in reply received by raft leader it
indicates that there is no consensus instance at the target node that
can handle the request. Logging this error explicitly instead of logging
false information about the target node being incorrect

Signed-off-by: Michal Maslanka <michal@vectorized.io>

mmaslankaprv committed:
c/errc: introduced invalid_configuration_update error code

Signed-off-by: Michal Maslanka <michal@vectorized.io>

mmaslankaprv committed:
c/members_manager: validating configuration update before applying

Introduced validation of configuration update request to prevent
situations in which two different redpanda nodes have the same RPC or
Kafka listener addresses.

Validation will prevent corruption of cluster brokers representation
which may lead to cluster being not usable.

Signed-off-by: Michal Maslanka <michal@vectorized.io>

mmaslankaprv committed:
r/errc: removed default case from raft errc print operator

Signed-off-by: Michal Maslanka <michal@vectorized.io>

VadimPlh committed:
Add executor for v8 engine

RafalKorepta committed:
Merge pull request #1701 from zimbatm/configurator-container-version

operator/helm: sync configurator image tag with operator tag

VadimPlh committed:
Add wrapper for v8_platfrom and add executor to env

BenPope committed:
schema_resgistry/storage: Support schema_key magic of 0

Magic 1 was introduced to support the deleted flag in schema_value.

The deserialiser implicitly supports both versions due to the default.

Signed-off-by: Ben Pope <ben@vectorized.io>

VadimPlh committed:
Add executor in script

0x5d committed:
Merge pull request #1699 from twmb/plugin_autocompletion

Plugin autocompletion

BenPope committed:
schema_registry: Fix swagger indentation

Signed-off-by: Ben Pope <ben@vectorized.io>

BenPope committed:
schema_registry: Introduce swagger for put_config

Signed-off-by: Ben Pope <ben@vectorized.io>

mmaslankaprv committed:
Merge pull request #1713 from mmaslankaprv/improved-raft-error-handling

Improved raft error handling

dotnwat committed:
Merge pull request #1632 from BenPope/schema-registry-config

Schema registry config

Commit:
rpk/acl: Simplify printing functions

Commit:
rpk acl list: Show the resource name pattern type

dotnwat committed:
Merge pull request #985 from 0x5d/rpk-ducky

testing: add ducktape tests for rpk

Lazin commented on src/v/v8_engine/executor.cc:
```diff
@@ -0,0 +1,134 @@
+
+ss::future<> executor::stop() {
+    return _gates.stop().then([this] {
+        _is_stopped = true;
```
неверный порядок действий при остановке, gate.stop будет ждать пока мы не выйдем из всех блоков with_gate, а потом уже установит флаг и вызовет notify
мне кажется нужно делать наоборот, установить флаг и вызвать notify, чтобы закончить ожидание на condition variable, а потом закрывать гейт

VadimPlh replied:
Кажется, что порядок верный
Сделал так, чтобы semaphore умел прерывать ожидание

Lazin commented on src/v/v8_engine/executor.cc:
```diff
@@ -0,0 +1,134 @@
+executor::executor(size_t threads_num, uint8_t cpu_id)
+  : _sem(threads_num)
+  , _tasks(threads_num) {
+    for (size_t i = 0; i < threads_num; i++) {
```
почему мы запускам несколько потоков, но передаем один cpu_id?

VadimPlh replied:
ЛОгика тредпула была в том, что он потоки к одномку ядру прицепляет. Я бы ее убрал, наверное

Lazin commented on src/v/v8_engine/executor.h:
```diff
@@ -0,0 +1,162 @@
+private:
+    std::atomic<size_t> _free_slots;
+
+    size_t MAX_WAIT_TIME_MS{100};
```
std::chrono::milliseconds вместо size_t?
или может быть сразу ss::lowres_clock::duration?

Lazin replied:
еще есть тип model::milliseconds

VadimPlh replied:
done

Lazin commented on src/v/v8_engine/executor.cc:
```diff
@@ -0,0 +1,134 @@
+                        wait_time *= 2;
+                    }
+
+                    return ss::make_ready_future<void>();
```
return ss::now();

VadimPlh replied:
готово

Lazin commented on src/v/v8_engine/executor.cc:
```diff
@@ -0,0 +1,134 @@
+          },
+          [this, &wait_time] {
+              return ss::sleep(std::chrono::milliseconds(wait_time))
+                .then([this, wait_time]() mutable {
```
wait_time копируется в эту лямбду, mutable просто позволяет изменять эту локальную копию, соотв. тебе нужно передавать в лямбду ссылку на wait_time, иначе она не будет меняться

еще тип size_t лучше заменить на duration или chrono::milliseconds
начальное значение 1 не имеет смысла, точность таймера 10мс, если передаешь 1мс, он ждет 0мс

VadimPlh replied:
Ой, я тут случафно забыл ссылку добавить, спаибо

VadimPlh replied:
готово

Lazin commented on src/v/v8_engine/script.h:
```diff
@@ -84,10 +88,29 @@ class script {
+    // Script timeout
+    size_t _timeout_ms;
+    // Timeout for run script in first time for initialization
+    static constexpr size_t FISRT_RUN_TIMEOUT_SEC = 10;
```
почему бы сразу не использовать тип для времени?
капс тоже не стоит использовать, мне кажется мы его нигде не используем

VadimPlh replied:
done

Lazin commented on src/v/v8_engine/script.h:
```diff
@@ -40,12 +42,13 @@ class script {
     // Init new instance.
     // Create isolate and configure constraints for it.
-    explicit script(size_t max_heap_size_in_bytes);
+    explicit script(
```
explicit не нужен

VadimPlh replied:
done

0x5d committed:
Merge pull request #1722 from 0x5d/rpk-acl-list-pattern

Rpk acl list pattern

Lazin committed:
s3: Recompute derived keys

AWS signature requires computation of the derived key which includes
the date component. This derived key expires after two days. Current
implementation assumes that the client is created per-request. This
is no longer the case after client pool was introduced to s3.

This commit adds ability to check the expiration date and recompute
the derived key if needed. The method should be called manually
before signing the header.

Lazin committed:
archival: Improve observability in case of error

Archival subsystem has the upload loop (UL) which is a fiber
that runs asynchronously for every ntp. It's responsibe for
uploading new segments and maintaining manifests. In case of
successful upload it starts upload next set of segments immediately.

If there is nothing to upload the loop uses exponential backoff
limiting max backoff interval to 5s. It start with pretty short
backoff interval of 10ms. This was done to prevent high CPU utilization
if there is no data to upload.

In case of error it's more complicated. The retriable errors (like
SlowDown response from S3) gets retried and backoff is applied
per request. This reties respect the total time budget of the
upload. But if the error is not retrieable (for instance NoSuchKey)
no retries are performed and the error is propagated.

After any error upload loop goes to the next iteration, picks the same
set of segments to upload and starts again. This leads to pretty
busy error output if the error is not retrieable.

This commit makes upload loop use exponential backoff if no progress
was made on the previous iteration.

Also, this commit discards the result of the failed future in the
reconciliation loop. The failure was handled but exceptional future
wasn't handled correctly which lead to warning being added to the log.

ivotron committed:
Merge pull request #1433 from VadimPlh/add-v8-env

Add first version for v8 instance and enviroment for v8

emaxerrno committed:
Merge pull request #1730 from Lazin/fix/archival-excessive-errors-and-signature-bug

archival: Fix upload errors and busy error logging
